arangod/Graph/Enumerators/WeightedTwoSidedEnumerator.h

```diff
@@ -59,18 +63,96 @@ template<class QueueType, class PathStoreType, class ProviderType,
+
+  // A meeting point with calculated path weight
+  using CalculatedCandidate = std::tuple<double, Step, Step>;
 
  private:
   enum Direction { FORWARD, BACKWARD };
 
   using VertexRef = arangodb::velocypack::HashedStringRef;
 
   using Shell = std::multiset<Step>;
-  using ResultList = std::vector<std::pair<Step, Step>>;
+  using ResultList = std::deque<CalculatedCandidate>;
   using GraphOptions = arangodb::graph::TwoSidedEnumeratorOptions;
 
+  /*
+   * Weighted candidate store for handling path-match candidates in the order
+   * from the lowest weight to the highest weight.
+   */
+  class CandidatesStore {
+   public:
+    void clear() {
+      if (!_queue.empty()) {
+        _queue.clear();
+      }
+    }
+
+    void append(CalculatedCandidate candidate) {
+      // if emplace() throws, no harm is done, and the memory usage increase
+      // will be rolled back
+      _queue.emplace_back(std::move(candidate));
+      // std::push_heap takes the last element in the queue, assumes that all
+      // other elements are in heap structure, and moves the last element into
+      // the correct position in the heap (incl. rebalancing of other elements)
+      // The heap structure guarantees that the first element in the queue
+      // is the "largest" element (in our case it is the smallest, as we
+      // inverted the comparator)
+      std::push_heap(_queue.begin(), _queue.end(), _cmpHeap);
+    }
+
+    size_t size() const { return _queue.size(); }
+
+    bool isEmpty() const { return _queue.empty(); }
+
+    std::vector<CalculatedCandidate> getQueue() const { return _queue; };
+
+    CalculatedCandidate& peek() {
+      TRI_ASSERT(!_queue.empty());
+      // will return a pointer to the step with the lowest weight amount
+      // possible. The heap structure guarantees that the first element in the
+      // queue is the "largest" element (in our case it is the smallest, as we
+      // inverted the comparator)
+      auto& first = _queue.front();
+      return first;
+    }
+
+    CalculatedCandidate pop() {
+      TRI_ASSERT(!isEmpty());
+      // std::pop_heap will move the front element (the one we would like to
+      // steal) to the back of the vector, keeping the tree intact otherwise.
+      // Now we steal the last element.
+      std::pop_heap(_queue.begin(), _queue.end(), _cmpHeap);
+      CalculatedCandidate first = std::move(_queue.back());
+      _queue.pop_back();
+      return first;
+    }
+
+   private:
+    struct WeightedComparator {
+      bool operator()(CalculatedCandidate const& a,
+                      CalculatedCandidate const& b) {
+        auto const& [weightA, candAA, candAB] = a;
+        auto const& [weightB, candBA, candBB] = b;
+        if (weightA == weightB) {
+          /*if (a.getDepth() == b.getDepth()) {
+            // Only false if A is not processable but B is.
+            return !a.isProcessable() || b.isProcessable();
+          }*/
+          return weightA > weightB;
```

Comment:
so now you first compare for equality in line 137, and then return whether `weightA > weightB`; The requirement for `cmp` is that it returns `true`  if the first argument is less than the second, so just `return weightA > weightB;` should be sufficient and correct here.

Reply:
lol! good that you've seen this. Changed to:

```
    struct WeightedComparator {
      bool operator()(CalculatedCandidate const& a,
                      CalculatedCandidate const& b) {
        auto const& [weightA, candAA, candAB] = a;
        auto const& [weightB, candBA, candBB] = b;
        return weightA > weightB;
      }
    };
```

Reply:
one more thing: in a std::tuple, can we do something like "std::ignore" here for candAA, candAB, etc. ?

Reply:
My preferred way to address this would be to not use a tuple ;)

you could just do 
```
    struct WeightedComparator {
      bool operator()(CalculatedCandidate const& a,
                      CalculatedCandidate const& b) {
        auto weightA = std::get<0>(a);
        auto weightB = std::get<0>(b);
        return weightA > weightB;
      }
    };
```

and I would *not* in this case remove the assignments to variables because then it becomes entirely unclear what the 0th element of the yple is. If the code is changed to using a `struct` instead of a tuple fo a candidate, then you'd have a member `weight` and could write `a.weight > b.weight`.